smarts/env/wrappers/parallel_env.py

```diff
@@ -0,0 +1,358 @@
+        """
+
+        self._assert_is_running()
+        if self._state != AsyncState.WAITING_STEP:
+            raise NoAsyncCallError(
+                "Calling `step_wait` without any prior call to `step_async`.",
+                AsyncState.WAITING_STEP.value,
+            )
+
+        if not self._poll(timeout):
+            self._state = AsyncState.DEFAULT
+            raise mp.TimeoutError(
+                "The call to `step_wait` has timed out after "
+                "{0} second{1}.".format(timeout, "s" if timeout > 1 else "")
+            )
+
+        results, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
+        self._raise_if_errors(successes)
+        self._state = AsyncState.DEFAULT
+        observations_list, rewards, dones, infos = zip(*results)
+
+        return (
+            observations_list,
+            rewards,
+            dones,
+            infos,
+        )
+
+    def _get_spaces(self) -> Tuple[gym.Space, gym.Space]:
+        for pipe in self.parent_pipes:
+            pipe.send(("_get_spaces", None))
+        spaces, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
+        self._raise_if_errors(successes)
+
+        observation_space = spaces[0][0]
+        action_space = spaces[0][1]
+
+        if not all([space[0] == observation_space for space in spaces]) or not all(
+            [space[1] == action_space for space in spaces]
+        ):
+            raise RuntimeError(
+                f"Expected all environments to have the same observation and action"
+                f"spaces, but got {spaces}."
+            )
+
+        return observation_space, action_space
+
+
+def _worker(
+    index: int,
+    sim_name: str,
+    env_constructor: CloudpickleWrapper,
+    auto_reset: bool,
+    pipe: mp.connection.Connection,
+    error_queue: mp.Queue,
+    polling_period: float = 0.1,
+):
+    """Process to build and run an environment. Using a pipe to
+    communicate with parent, the process receives action, steps
+    the environment, and returns the observations.
+
+    Args:
+        index (int): Environment index number.
+        env_constructor (CloudpickleWrapper): Callable which constructs the environment.
+        auto_reset (bool): If True, auto resets environment when episode ends.
+        pipe (mp.connection.Connection): Child's end of the pipe.
+        error_queue (mp.Queue): Queue to communicate error messages.
+        polling_period (float): Time to wait for keyboard interrupts.
+    """
+
+    # Name and construct the environment
+    name = f"env_{index}"
+    if sim_name:
+        name = sim_name + "_" + name
+    env = env_constructor(sim_name=name)
+
+    # Environment setup complete
+    pipe.send((None, True))
+
+    try:
+        while True:
+            # Short block for keyboard interrupts
+            if not pipe.poll(polling_period):
+                continue
```

Comment:
How does this raise keyboard interrupts that are not seen by any of the other `Pipe` methods?

Reply:
```python
# Short block for keyboard interrupts
if not pipe.poll(polling_period):
    continue
```

This is to make the code more responsive to a `CTRL-C` command.